app-scaling/README.adoc

```diff
@@ -0,0 +1,74 @@
+
+= Kubernetes - App Scaling
+:icons:
+:linkcss:
+:imagesdir: ../images
+
+== Kubernetes - App Scaling
+
+Horizontal Pod Autoscaling (HPA) is a Kubernetes feature to dynamically increase/decrease the number of pod replicas based on resource utilization metrics.
+
+Heapster is an open source project that collects cluster metrics, and Kubernetes integrates with it to determine CPU consumption. Heapster's default backend, referred to as a "sink", is InfluxDB. Additional sinks are supported including Elasticsearch.
+
+Kubernetes currently has beta support for custom metrics from multiple sources, and not just Heapster. This is enabled via the API Server Aggregation feature, and this documentation will be updated when the feature is live to show HPA with additional types of metrics.
+
+HPA can automatically scale pods deployed in a replication controller, deployment, or a replica set. For additional information on how HPA works, check out the Kubernetes https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/[community documentation].
+
+=== Prerequisistes
+
+1. https://github.com/arun-gupta/kubernetes-aws-workshop/tree/master/install-cluster[Deploy a Kubernetes cluster] and verify that it works
+2. Deploy Heapster using the steps described in the https://github.com/arun-gupta/kubernetes-aws-workshop/tree/master/cluster-monitoring[Cluster Monitoring] lab.
+
+=== Deploy an Application
+
+In this step, we deploy a simple Go web application and constrain the CPU resources just for the purposes of this test.
+
+    $ kubectl run webapp --image=trevorrobertsjr/webapp --requests=cpu=50m --expose --port=8080
+
+=== Horizontal Pod Autoscaler Configuration
+
+Now that our application is running, we create a Horizonal Pod Autoscaler for our webapp deployment.
+
+    $ kubectl autoscale deployment webapp --cpu-percent=10 --min=1 --max=10
+
+=== Generate Load
+
+The simplest method to do this would be to access the application in an infinite loop similar to the example in the Kubernetes Horizonal Pod Autoscaler documentation:
+
+First, deploy a busybox container, label it *load-generator* and attach to it's prompt:
+
+    $ kubectl run -i --tty load-generator --image=busybox /bin/sh
+
+At the *load-generator*'s command prompt, run a continuous request of the webapp
+
+    $ while true; do wget -q -O- http://webapp.default.svc.cluster.local; done
+
+If for any reason you get disconnected from the load-generator container, you can re-attach to it with the following command.
+
+    $ kubectl attach $(kubectl get pod | grep load | awk '{print $1}') -c load-generator -i -t
+
+In a different terminal window, check the status of the Horizontal Pod Autoscaler.
+
+    $ kubectl get hpa
+
+You will see output similar to the following over successive executions of the *kubectl get hpa* command:
```

Comment:
Might make sense to show off `-w` so something like `kubectl get hpa -w` so they don't have to subsequently run it multiple times. Thoughts? @TrevorRobertsJr

Reply:
Hello @christopherhein I did not see a notification for your comment. That makes sense. I will update accordingly.